File: frontend/src/types/holding.ts
export interface Holding {
  id: number;
  ticker: string;
  company_name: string | null;
  quantity: number;
  average_buy_price: number;
  current_price: number;
  total_invested: number;
  current_value: number;
  unrealized_gain: number;
  unrealized_gain_pct: number;
  allocation_pct: number;
  sector: string | null;
  industry: string | null;
  dividend_yield: number;
  annual_dividend_income: number;
  broker: string | null;
  created_at: string;
  updated_at: string;
}

export interface HoldingCreate {
  ticker: string;
  quantity: number;
  buy_price: number;
  company_name?: string;
  sector?: string;
  industry?: string;
  broker?: string;
  notes?: string;
}

File: backend/app/schemas/holding.py
"""Pydantic schemas for holding-related requests and responses."""

from datetime import datetime

from pydantic import BaseModel, Field


class HoldingCreate(BaseModel):
    """Schema for creating a new holding."""

    ticker: str = Field(..., min_length=1, max_length=5, pattern=r"^[A-Z]{1,5}$")
    quantity: float = Field(..., gt=0)
    buy_price: float = Field(..., ge=0.01)
    company_name: str | None = None
    sector: str | None = None
    industry: str | None = None
    broker: str | None = None
    notes: str | None = None


class HoldingUpdate(BaseModel):
    """Schema for updating an existing holding."""

    company_name: str | None = None
    sector: str | None = None
    industry: str | None = None
    notes: str | None = None


class HoldingResponse(BaseModel):
    """Schema for holding response data."""

    id: int
    ticker: str
    company_name: str | None
    quantity: float
    average_buy_price: float
    current_price: float
    total_invested: float
    current_value: float
    unrealized_gain: float
    unrealized_gain_pct: float
    allocation_pct: float
    sector: str | None
    industry: str | None
    dividend_yield: float
    annual_dividend_income: float
    broker: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
